src/Core/src/Fonts/EmbeddedFontLoader.Windows.cs

```diff
@@ -34,7 +39,7 @@ public partial class EmbeddedFontLoader
 			}
 			catch (Exception ex)
 			{
-				_logger?.LogWarning(ex, "Unable copy font {Font} to local file system.", font.FontName);
+				_serviceProvider?.CreateLogger<FontManager>()?.LogWarning(ex, "Unable copy font {Font} to local file system.", font.FontName);
```

Comment:
Can `_serviceProvider` be non-nullable and readonly in each of these classes? So we wouldn't have to do `_serviceProvider?.Something()` when using it?

Reply:
Even though these are public, I assume we don't expect "normal" callers to create them? They should get them through DI, right?

Reply:
This may no longer be true, but is ctor arg injection slower than just this func? It has to build up and search overloads and things.

Or is the difference negligible?

Reply:
There are places that explicitly pass `null` in for these parameters. For example:

https://github.com/dotnet/maui/blob/ec33802c4ab453e13a24b1fc63bb019cd28152f9/src/Core/src/Fonts/EmbeddedFontLoader.cs#L20-L21

Reply:
Issue with iserviceprovider args is that it does not describe what you are looking for. Maybe we should just use the overload that takes a func?

Reply:
> This may no longer be true, but is ctor arg injection slower than just this func? It has to build up and search overloads and things.

In the case of Logging, initializing a Logger is what is slow. I can't really make it faster without redesigning Logging all up - there is just so many layers to it, and the fact that it uses generic types, and that its constructor uses a StringBuilder to build up a category name, etc. It is just a really slow thing to create. So if we don't need it, don't create it.

If you use ctor arg injection, then DI has to create your dependencies when it creates you. So there is no way to "lazy load" them this way.

> Maybe we should just use the overload that takes a func?

Can you explain this more?

Reply:
Use Func<ILogger<T>> in the args. It can be an overload.

Reply:
I'm not sure that would be much of an improvement. It would just be another layer of indirection, wouldn't it?

Instead of:

```C#
builder.Services.TryAddSingleton<IEmbeddedFontLoader>(svc => new EmbeddedFontLoader(svc));
```

it would be

```C#
builder.Services.TryAddSingleton<IEmbeddedFontLoader>(svc => new EmbeddedFontLoader(() => svc.CreateLogger<FontManager>()));
```

And we would still need to check it for `null`. So at the log site we would have:

```C#
_getLogger?.Invoke()?.LogWarning(ex, "Unable copy font {Font} to local file system.", font.FontName);
```

Reply:
I mean we could cache it on first use. This way the type is clear and also it is cached. Maybe there is a `Logger` property? In fact, we _should_ cache the logger, right? 

And if we always have a `NullLogger`, then this will never be null. But, I can't imagine that a null check is the real perf hit here.

Reply:
Basically, my main concern is that it takes `IServiceProvider` and that is a bit of a no-no in the DI world, but also it doesn't let the callers and mocking system know what to provide. Maybe we should have an overload - so both the type and the ISP

Right now, I don't think our tests actually check to see if there was an internal exception, and one way is the logger.

I am just scared of the opaque ISP... Maybe I am just pessimistic. If nobody else shares this concern, then maybe it is not a real issue and I am just a scared guy XD

Reply:
> is a bit of a no-no in the DI world, but also it doesn't let the callers and mocking system know what to provide. Maybe we should have an overload - so both the type and the ISP

I'd say it's a no-no with exceptions and that lot of that advice originated from scenarios that weren't trying to make an Android app start up quickly. 

As long as we don't start resolving 10s of dependencies through service locator then it's fine. 

If we want to have clear dependencies on the `FontManager` maybe we could register a singleton `IMauiLoggerFactory` and the concrete implementation of that takes an `IServiceProvider` on the `ctor`. Then the fuzziness of ISP is isolated to the factory implementation? Though maybe that much indirection will just start to ruin the benefits here.

Reply:
Just so I fully understand the concerns here, are the concerns with taking IServiceProvider here basically the same that are listed in:

https://blog.ploeh.dk/2015/10/26/service-locator-violates-encapsulation/
https://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/